gradle/forbidden-dependencies.gradle

```diff
@@ -13,7 +13,8 @@
 // we do not want any of these dependencies on the compilation classpath
 // because they could then be used within OpenSearch
 List<String> FORBIDDEN_DEPENDENCIES = [
-  'guava'
+  // TODO: Why is guava forbidden?
```

Comment:
I believe the answer is that in the past using guava has led to jar hell as all plugins depend on the `:server` module, and if guava comes as a transitive dependency then all plugins will be forced to use the same version of guava.

I think the answer here is to try to confine the gRPC dependency to a module (i.e. `module/transport-netty4`, or create a new `module/grpc`). I think this should be possible, with the `:server` module containing the logic for converting a proto request into a (streaming) proto response, but without explicit gRPC types. The module would then be responsible for wiring up this function to the correct gRPC-generated type.

Reply:
Actually, I'm pretty sure we should be able to keep most of the gRPC logic inside a plugin or module itself, as the dependencies needed by the gRPC layer should be accessible via the [`createComponents()`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/Plugin.java#L142-L156) injection method used by plugins/modules. A question remains whether to use the same Netty server as the `transport-netty4` module (which might require building this inside that module directly) versus using its own (probably the shaded version inside the gRPC library, which would probably make upgrading gRPC in the future easier).

The Guava issue is a bit trickier. The current logic just forbids Guava to be on a compile classpath anywhere inside this repo regardless of component. The gRPC-generated code requires Guava to compile. The only paths forward here I see are we add an exemption to allow Guava on the compile classpath for the component containing the generated code, or the generated code is compiled in an external repository that is brought in here as a dependency (having guava on the runtime classpath is fine).

Reply:
> A question remains whether to use the same Netty server as the transport-netty4 module (which might require building this inside that module directly) versus using its own (probably the shaded version inside the gRPC library, which would probably make upgrading gRPC in the future easier).

@andrross just for sharing https://github.com/grpc/grpc-java/issues/10931#issuecomment-1954901761. 
i guess the key difference is if we want to have isolation from netty conflicts vs we want to manage netty versions explicitly. With HTTP + GRPC co-exist, probably latter? Besides that, i don't see performance etc difference based on initial reading.

Reply:
> generated code is compiled in an external repository that is brought in here as a dependency (having guava on the runtime classpath is fine).

I think having the protobufs generated and imported as a jar/dependency makes more sense. 

1. It would also make it easier for other non-opensearch-core repos to use these protobufs down the line, without importing it as a submodule. For example the opensearch-java client, opensearch-benchmark, etc. 

2. During our implementation for the Search/Document endpoints, we found that other folders will also require protobuf files, including [libs/core/src/main/java/org/opensearch/OpenSearchException.java](https://github.com/opensearch-project/OpenSearch/blob/main/libs/core/src/main/java/org/opensearch/OpenSearchException.java). We cannot import the protos into this folder, as it would introduce a circular dependency between Opensearch/libs <> Opensearch/server. So putting the protobufs inside of `OpenSearch/server/` may not be the best choice.

Reply:
I think it would be great to decouple this problem from this PR. Per the other [comment](https://github.com/opensearch-project/OpenSearch/pull/16534#issuecomment-2521766761) down below, I think it would make sense to remove Protos and API implementations from this diff so we can get the base server merged first. What are your thoughts?